File: vercel-ai/tour-of-agents/src/workflow-parallel.ts
import * as restate from "@restatedev/restate-sdk";
import { openai } from "@ai-sdk/openai";
import { generateText, wrapLanguageModel } from "ai";
import {ClaimInput, ClaimInputSchema, ClaimPromptSchema} from "./utils/types";
import {
  eligibilityAgent,
  fraudCheckAgent,
  rateComparisonAgent,
} from "./utils/utils";
import { RestatePromise } from "@restatedev/restate-sdk";
import { durableCalls } from "@restatedev/vercel-ai-middleware";
const schema = restate.serde.schema;

const run = async (ctx: restate.Context, claim: ClaimInput) => {
  const [eligibility, rateComparison, fraudCheck] =
    await RestatePromise.all([
      ctx.serviceClient(eligibilityAgent).run(claim),
      ctx.serviceClient(rateComparisonAgent).run(claim),
      ctx.serviceClient(fraudCheckAgent).run(claim),
    ]);

  const model = wrapLanguageModel({
    model: openai("gpt-4o"),
    middleware: durableCalls(ctx, { maxRetryAttempts: 3 }),
  });

  const { text } = await generateText({
    model,
    system: "You are a claim decision engine.",
    prompt: `Decide about claim ${JSON.stringify(claim)}.
        Base your decision on the following analyses:
        Eligibility: ${eligibility}, Cost: ${rateComparison} Fraud: ${fraudCheck}`,
  });
  return text;
};

const agent = restate.service({
  name: "ParallelAgentClaimApproval",
  handlers: {
    run: restate.createServiceHandler({ input: schema(ClaimInputSchema) }, run),
  },
});

restate.serve({
  services: [agent, eligibilityAgent, rateComparisonAgent, fraudCheckAgent],
});

File: langchain-python/tour-of-agents/app/workflow_parallel.py
"""Run several specialist agents in parallel, then feed their answers into
a decision agent."""

import restate
from langchain.agents import create_agent
from langchain.chat_models import init_chat_model

from restate.ext.langchain import RestateMiddleware

from utils.models import InsuranceClaim
from utils.utils import (
    eligibility_agent_service,
    fraud_agent_service,
    rate_comparison_agent_service,
    run_eligibility_agent,
    run_fraud_agent,
    run_rate_comparison_agent,
)

# <start_here>
decision = create_agent(
    model=init_chat_model("openai:gpt-5.4"),
    system_prompt="You are a claim decision engine.",
    middleware=[RestateMiddleware()],
)


agent_service = restate.Service("ParallelAgentClaimApproval")


@agent_service.handler()
async def run(ctx: restate.Context, claim: InsuranceClaim) -> str:
    # Start multiple sub-agents in parallel with auto-retries and recovery.
    eligibility = ctx.service_call(run_eligibility_agent, claim)
    cost = ctx.service_call(run_rate_comparison_agent, claim)
    fraud = ctx.service_call(run_fraud_agent, claim)

    await restate.gather(eligibility, cost, fraud)

    result = await decision.ainvoke(
        {"messages": f"""Decide about claim: {claim.model_dump_json()}.
        Base your decision on the following analyses:
        Eligibility: {await eligibility} Cost {await cost} Fraud: {await fraud}"""}
    )
    return result["messages"][-1].content


# <end_here>


if __name__ == "__main__":
    import asyncio

    import hypercorn
    import hypercorn.asyncio

    app = restate.app(
        services=[
            agent_service,
            fraud_agent_service,
            rate_comparison_agent_service,
            eligibility_agent_service,
        ]
    )
    conf = hypercorn.Config()
    conf.bind = ["0.0.0.0:9080"]
    asyncio.run(hypercorn.asyncio.serve(app, conf))
